README
README
======

Biblio@Tech Net::FTPServer - A full-featured, secure, extensible
and configurable Perl FTP server.

The server was written as part of the Biblio@Tech Schoolmaster.net
project and was kindly released under the GNU General Public License
(GPL). 

Major features of Net::FTPServer include:

* Feature parity with wu-ftpd.
* IP-based and IP-less virtual hosting.
* Virtual filesystem allows the FTP server to serve files from a
  SQL database.
* Configurable and extensible in Perl.
* <Perl> sections in ftpd.conf file.
* Supports all the latest RFCs and Internet Drafts, including MLST,
  MLSD, FEAT, OPTS, LANG.
* Secure by design and implementation.
* PAM authentication.
* Resource limits.
* Run standalone or from inetd.
* Configurable server greetings and welcome messages.
* Anonymous mode.
* Run in chroot jail.
* Sophisticated access control rules.
* wu-ftpd style aliases and cdpath.
* SITE EXEC (disabled by default).
* Syslog logging.
* Set TCP parameters.

Missing features:

* Upload/download quotas.


PREREQUISITES
-------------
required:  perl >= 5.00503  IO               This is also normally included in the                   basic Perl package, but make sure you                   have a pretty recent version.  IO::stringy      This is now required by the main server.                   Make sure you have at least version 1.220                   from CPAN which includes some important                   bug fixes.optional:  Archive::Zip     If you want to allow users to create ZIP                   files from directories on the fly (so-called                   "archive mode"), you need this module.  Authen::PAM      If you want to do PAM authentication (eg.                   you have Red Hat Linux or Solaris).  BSD::Resource    If you want resource limits to avoid certain                   types of denial of service (DoS) attack.  Compress::Zlib   If you want to allow users to create ZIP                   files from directories on the fly (so-called                   "archive mode"), you need this module.  Digest::MD5      If you want the SITE CHECKSUM command                   to work correctly.  File::Sync       Needed for SITE SYNC command.  bzip2            If an external "bzip2" program is available,                   it can be used to create archives                   automatically.  gzip             If an external "gzip" program is available,                   it can be used to create archives                   automatically.  uuencode         If an external "uuencode" program is                   available, it can be used to create                   uuencoded archives automatically.INSTALLATION
------------
To compile:	perl Makefile.PL	make	make testTo install (usually as root):	make installYou can override where the configuration file will goby specifying the SYSCONFDIR setting.  /etc is the default.Or to avoid installing a configuration altogether use"make NOCONF=1 install" instead.The documentation should be automatically installed. Do:	man Net::FTPServeror if you don't have a working man pages system, do:	pod2man lib/Net/FTPServer.pm | nroff -man | moreRead the section "INSTALLING AND RUNNING THE SERVER" to find out howto change the configuration file to suit your local situation, and howto start and stop the server.CONFIGURATION AND DOCUMENTATION
-------------------------------

All other documentation is in the Net::FTPServer(3) man page. After
installation, type:

	man Net::FTPServer

and read the section "INSTALLING AND RUNNING THE SERVER".


SUPPORT AND MAILING LIST
------------------------

NOTE: the information below is still avtive, but please consider to
contact discussion forum on CPAN and/or github repositry first. 

There is a mailing list for general discussion and announcements.  See
the Savannah mailing list page for details:

https://savannah.nongnu.org/mail/?group=netftpserver

Bug reports, feature requests, and so on should be sent to the mailing
list.


CREDITS
-------

Copyright (C) 2000 Biblio@Tech Ltd., Unit 2-3, 50 Carnwath Road, London, SW6 3EG, UK.
Copyright (C) 2000-2003 Richard Jones <rich@annexia.org> and other contributors.


Changes
Changes
=======

Version 1.124
-------------

Thu Nov 22 13:26:09 UTC 2012

IMPROVEMENTS:

* addressed an issue that ftpd.conf had been installed despite lack of write permission to sysconfdir. [#81130]
  (thanks to Gabor Szabo)
* added a message that Win32 platform is not supported. [#81136]
  (thanks to CHORNY)
* switched to Test::More for better installation process.

Version 1.123
-------------

Tue Nov 13 03:08:08 UTC 2012    Maintenance release after a long time

BUG FIXES:

* added a workaround to make sure to cause abort by SIGURG in t/240abort.t. [#21261]
* fixed a bug that MLST command treated DirHandle as FileHandle. [#27640]
  (thanks to Andrew)
* fixed a bug that ls -l command doesn't show a directory named '0'. [#29503]
* fixed a problem that extra large file sizes were not displayed correctly because of integer digit overflow. [#35332]
  (thanks to Kevin Morga)
* fixed a problem caused by a Constant.pm internal change, affected to Archive::Zip. [#35698]
  (thanks to Alexander Bluhm)
* fixed a problem that Archive::Zip::Member::setLastModFileDateTimeFromUnix() doesn't accept 0. [#35698]
  (thanks to Alexander Bluhm)

OTHERS:

* maintainer now changed.
* organized document and package files.
* fixed pod format errors.
* new repository: https://github.com/ryochin/p5-net-ftpserver

Version 1.122
-------------

(No description)

Version 1.121
-------------

(No description)

Version 1.120
-------------

Added PREREQ_PM to Makefile.PL
so CPAN works (thanks to barbie at missbarbell dot co dot uk).  Fixed
a few warnings (thanks David Wheeler, david at kineticode dot com).

Version 1.119
-------------

Added FTP proxy backend (highly
experimental - please test). Patch from John Jetmore to open
/etc/localtime in chroot environment.  All of the following from Yair
Lenga: 'rotate log files' configuration.  open_data_connection fix for
non-reusable sockets on SunOS.  'die' replaced with 'confess' in a few
places to aid debugging.  Allow 'ftp data port' to be given as a
service name, as well as a number.  'authentication_hook' may return
-2 to avoid sleep.  Don't drop priviledges if 'ftp data port' is bound
to a socket < 1024.  Send back GMT in MDTM command.  Various extra
directives such as '%Y' can be included in log file names
(undocumented?).

Version 1.118
-------------

Moved to Savannah.nongnu.org.

Version 1.117
-------------

Fix t/350filters.t uudecode
syntax to comply with certain systems.

Version 1.116
-------------

Incorporated a change from the
FreeBSD port of Net::FTPServer - does not install with '-g root'
(which is a Linux-ism). Thanks for petef at freebsd.org for the patch.
Less strict regex for PORT command (spot by Brian.Freeman at
eby-brown.com). Allow "make test" to pass when certain filter commands
are missing. Fix t/350filters.t compress syntax to comply with
certain systems (spot by Vicki Brown  vlb at cfcl.com). Allow name
based port to be used for source port on the server endpoint for PORT
mode transfers. (i.e., "ftp data port: ftp-data" works now.  This is
required by some router/firewall configurations. - Rob Brown)
Compatibility fixes for ExtUtils::MakeMaker (5.91, 6.03, 6.10_03).
Other minor package/build/make/rpm/install changes.

Version 1.115
-------------

Remove bootstrap code from spec
file so rpm works again. (Broken since 1.112.)  Add "ftp data port"
directive to allow server to comply closer with RFC 959 if the user
wishes to do so.  But for security reasons, the default remains to use
the high ephemeral port.  -- Rob

Version 1.114
-------------

No new features. Bug fixes only.
Fix t/150commands.t LANG test failure on some systems.  Avoid race
condition in t/400sighup.t that can occur when client connects before
SIGHUP is done restarting the server.  Fixed the RMD bug introduced
from Luis' Darwin / Mac OS X "removing directoring" patch in 1.112.
Now RMD works again for the "Full" personality.  -- Rob

Version 1.113
-------------

Removed all XS code. Signal
handlers are pure perl now even for older versions of perl < 5.7.
Concept taken from Net::Server::SIG in order to avoid dangerous
pitfalls of unsafe signal handling in older versions of perl. -- Rob

Version 1.112
-------------

Avoids sleeping in Makefile.PL if
you set NET_FTPSERVER_NO_SLEEP=1. If Perl version >= 5.7.2, then the
XS-based safe signals code is disabled. These later Perl versions
include safe signals, so it is not needed. Note to package
maintainers: if using Perl >= 5.7.2, then Net::FTPServer is pure Perl
code, no C. Allow non-alpha characters in SITE commands (feature
requested by Brian Freeman). Luis Mun\~oz fixed a problem with
removing directories on Darwin / Mac OS X. Luis also added the 'chdir
rule' which allows finer control over what directories a user may
change into. Allow the build to work if there is a different version
of the server installed.  Changes by Rob Brown: 1) Reduce redundant
code.  2) Allow easier overrides for Full and RO personalities.  (That
whole "sub list" doesn't need to be copied.)  3) Avoid spewing
warnings to the logs and showing garbage to the client for a race
condition where the node has been removed between the time of the
IO::Dir read and the time of the @array stashing.

Version 1.111
-------------

Fixed a critical memory
corruption error. Fixed and removed a faulty call to 'fcntl' which
passed a parameter by value where a pointer was expected. (One symptom
was that messages produced by the '-d' option in the system log could
appear blank). Work around a bug in Perl 5.6.0 which was causing the
SYST command (and possibly other commands) to fail at random because a
regular expression match gave $1 a corrupted value.

Version 1.110
-------------

Workaround for a bug in
IE 6.2600.0000. This version of IE sends a leading 0 before
the IP address in the PORT command (eg. PORT 0212,57,32,3,1,2).

Version 1.109
-------------

Tested the FTP server on Perl
5.00405, so changed the Makefile.PL to allow this version. (Note on

Makefile.PL
# -*- perl -*-

# Net::FTPServer A Perl FTP Server
# Copyright (C) 2000 Bibliotech Ltd., Unit 2-3, 50 Carnwath Road,
# London, SW6 3EG, United Kingdom.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

use ExtUtils::MakeMaker;

# PREREQ_PM is crap! It doesn't enforce the prerequisites, and
# doesn't print an intelligible error message. Nor does it understand
# that some modules are optional, while others are absolutely
# required. Do our own prerequisite checking here. A lot of this
# code was borrowed from the Makefile.PL supplied with libwww-perl.
#
# NB. 'status' field is either 'required', for modules which are
# required, or anything else for modules which are optional.

# OS check
if( $^O eq 'MSWin32' ){
    printf STDERR "This module doesn't support your platform $^O.\n";
    exit 1;
}

my %modules
  = (
     'Archive::Zip' =>
     { status => "optional",
       version => "0.11",
       message =>
       "Archive::Zip is missing. This module is required if you want to\n".
       "enable archive mode, which allows users to create ZIP files on\n".
       "the fly from directories on the server.\n"
     },
     'Authen::PAM' =>
     { status => "recommended",
       version => "0.12",
       message =>
       "Authen::PAM is missing. This module is required in order to\n".
       "perform user authentication against the full FTP server\n".
       "personality on most Unix platforms. Therefore unless you install\n".
       "this module, normal FTP server configurations will most likely\n".
       "not work.\n"
     },
     'BSD::Resource' =>
     { status => "highly recommended",
       message =>
       "BSD::Resource is missing. This module is required in order to\n".
       "place limits on the size of the FTP server process. Without this\n".
       "module, the FTP server will still work, but it will be vulnerable\n".
       "to certain sorts of denial of service (DoS) attacks.\n"
     },
     'Carp' => { status => "required" },
     'Compress::Zlib' =>
     { status => "optional",
       version => "1.14",
       message =>
       "Compress::Zlib is missing. This module is required if you want to\n".
       "enable archive mode, which allows users to create ZIP files on\n".
       "the fly from directories on the server.\n"
     },
     'DBI' =>
     { status => "optional",
       message =>
       "DBI is missing. if you want to run the FTP server against a\n".
       "PostgreSQL database.\n"
     },
     'Digest::MD5' =>
     { status => "optional",
       message =>
       "Digest::MD5 is missing. This module is required in order for the\n".
       "SITE CHECKSUM command to work.\n"
     },
     'DynaLoader' => { status => "required" },
     'Fcntl' => { status => "required" },
     'File::Sync' =>
     { status => "optional",
       message =>
       "File::Sync is missing. The SITE SYNC command will not work.\n"
     },
     'File::Temp' =>
     { status => "option",
       message =>
       "File::Temp is missing. This is required for the proxy server.\n"
     },
     'Getopt::Long' => { status => "required" },
     'IO::Dir' => { status => "required", package => "IO" },
     'IO::File' => { status => "required", package => "IO" },
     'IO::Handle' => { status => "required", package => "IO" },
     'IO::Scalar' => { status => "required", package => "IO-stringy",
		       version => "1.126" },
     'IO::Seekable' => { status => "required" },
     'IO::Select' => { status => "required", package => "IO" },
     'IO::Socket' => { status => "required", package => "IO" },
     'IPC::Open2' => { status => "required" },
     'Net::FTP' =>
     { status => "optional",
       message =>
       "Net::FTP is missing. This is required for the proxy server.\n"
     },
     'POSIX' => { status => "required" },
     'Socket' => { status => "required" },
     'Sys::Hostname' => { status => "required" },
     'Sys::Syslog' => { status => "required" },
     'Test::More' => { status => "required" },
    );

$| = 1;

# Check for modules.

my $missing_modules = 0;
my $missing_required_modules = 0;

foreach (sort keys %modules)
  {
    print "Checking for $modules{$_}{status} ";
    print "module $_ ";
    print ">= $modules{$_}{version} " if $modules{$_}{version};
    print "... ";

    my $eval = "require $_; ";
    $eval .= "$_->VERSION >= $modules{$_}{version}" if $modules{$_}{version};

    my $r = eval $eval;
    if ($@ || !$r)
      {
	$missing_modules++;
	$missing_required_modules++ if $modules{$_}{status} eq "required";

	print "not found.\n\n";

	if (exists $modules{$_}{package})
	  {
	    print
	      "This module is provided by the $modules{$_}{package} ",
	      "package.\n\n";
	  }

	sleep 1 unless $ENV{NET_FTPSERVER_NO_SLEEP};

	if (exists $modules{$_}{message})
	  {
	    print "*** ", $modules{$_}{message}, "\n";
	    sleep 5 unless $ENV{NET_FTPSERVER_NO_SLEEP};
	  }

      }
    else
      {
	print "ok.\n";
      }
  }

print "\n";

if ($missing_modules)
  {
    print <<EOT;
Obtain missing modules from CPAN [http://www.cpan.org/].

EOT
    sleep 2 unless $ENV{NET_FTPSERVER_NO_SLEEP};
  }

if ($missing_required_modules)
  {
    print <<EOT;
Required modules are missing. Install these first.

EOT
    exit 1;
  }

# The code has only been verified on versions of Perl >= than
# the following. If you find an earlier version of Perl on which
# Net::FTPServer runs, please tell me. - rich@annexia.org.
if ($] < 5.00405)
  {
    print <<EOT;
Perl version >= 5.00405 is required.

EOT
    exit 1;
  }

my @args = (
	    NAME => "Net::FTPServer",
	    VERSION_FROM => "lib/Net/FTPServer.pm",

	    # Install start-up scripts under /usr/sbin.
	    EXE_FILES => [ 'bin/dbeg1-ftpd.pl',
			   'bin/ftpd.pl',
			   'bin/inmem-ftpd.pl',
			   'bin/proxy-ftpd.pl',
			   'bin/ro-ftpd.pl' ],
	    INSTALLSCRIPT => $ExtUtils::MakeMaker::VERSION < 5.45 ?
	    '$(PREFIX)/sbin' : # No SITEPREFIX before Perl 5.6.0
	    '$(SITEPREFIX)/sbin',

	    dist =>
	    {
	     COMPRESS => 'gzip --force --best',
#	     PREOP => 'cp Net-FTPServer.spec $(DISTVNAME)',
	    },

	    clean =>
	    {
	     FILES => '*.bak *~',
	    },

	    realclean =>
	    {
#	     FILES => 'Net-FTPServer.spec',
	    },

	    depend =>
	    {
#	     dist => 'Net-FTPServer.spec',
	     #'Net-FTPServer.spec' => "lib/Net/FTPServer.pm",
	     #Makefile => "lib/Net/FTPServer.pm",
	    },

	    # Despite what I say about PREREQ_PM above, we need it to
	    # satisfy various automatic tools, eg. CPAN.
	    PREREQ_PM   =>
	    {
	     'BSD::Resource'    => '0',
	     'Carp'		=> '0',
	     'DynaLoader'	=> '0',
	     'Fcntl'		=> '0',
	     'Getopt::Long'	=> '0',
	     'IO::Dir'		=> '0',
	     'IO::File'		=> '0',
	     'IO::Handle'	=> '0',
	     'IO::Scalar'	=> 1.126,
	     'IO::Seekable'	=> '0',
	     'IO::Select'	=> '0',
	     'IO::Socket'	=> '0',
	     'IPC::Open2'	=> '0',
	     'POSIX'		=> '0',
	     'Socket'		=> '0',
	     'Sys::Hostname'	=> '0',
	     'Sys::Syslog'	=> '0',
	    },
	    BUILD_REQUIRES => {
	      'Test::More'	=> '0',
	    },

	    LICENSE => 'gpl',
	    MIN_PERL_VERSION => '5.005',
	    META_MERGE => {
	      resources => {
	        repository => 'https://github.com/ryochin/p5-net-ftpserver',
	      },
	    },
);

WriteMakefile1(@args);

sub WriteMakefile1 {  #Compatibility code for old versions of EU::MM. Written by Alexandr Ciornii, version 0.23. Added by eumm-upgrade.
  my %params=@_;
  my $eumm_version=$ExtUtils::MakeMaker::VERSION;
  $eumm_version=eval $eumm_version;
  die "EXTRA_META is deprecated" if exists $params{EXTRA_META};
  die "License not specified" if not exists $params{LICENSE};
  if ($params{AUTHOR} and ref($params{AUTHOR}) eq 'ARRAY' and $eumm_version < 6.58) {
    $params{META_ADD}->{author}=$params{AUTHOR};
    $params{AUTHOR}=join(', ',@{$params{AUTHOR}});
  }
  if ($params{BUILD_REQUIRES} and $eumm_version < 6.5503) {
    #EUMM 6.5502 has problems with BUILD_REQUIRES
    $params{PREREQ_PM}={ %{$params{PREREQ_PM} || {}} , %{$params{BUILD_REQUIRES}} };
    delete $params{BUILD_REQUIRES};
  }
  delete $params{CONFIGURE_REQUIRES} if $eumm_version < 6.52;
  delete $params{MIN_PERL_VERSION} if $eumm_version < 6.48;
  delete $params{META_MERGE} if $eumm_version < 6.46;
  delete $params{META_ADD} if $eumm_version < 6.46;
  delete $params{LICENSE} if $eumm_version < 6.31;
  delete $params{AUTHOR} if $] < 5.005;
  delete $params{ABSTRACT_FROM} if $] < 5.005;
  delete $params{BINARY_LOCATION} if $] < 5.005;

  WriteMakefile(%params);
}


package MY;

sub processPL
  {
    my $self = shift;
    my $block = $self->SUPER::processPL(@_);
    # "Version:" in spec needs to match
    # "$VERSION" from VERSION_FROM
    $block =~ s%(spec.PL\s*)$%$1 \$\(VERSION_FROM\)%m;
    $block;
  }

sub libscan
  {
    my ($self, $path) = @_;
    ($path =~ /\~$/) ? undef : $path;
  }

sub postamble
  {
    '
sysconfdir	= $(DESTDIR)/etc

check-manifest:
	@for d in `find -type d -name CVS`; \
	do \
	b=`dirname $$d`/; \
	awk -F/ \'$$1 != "D" {print $$2}\' $$d/Entries | \
	sed -e "s|^|$$b|" -e "s|^\./||"; \
	done | sort > .check-manifest; \
	sort MANIFEST > .orig-manifest; \
	diff -u .orig-manifest .check-manifest; \
	rm -f .orig-manifest .check-manifest

install ::
	@if [ ! "${NOCONF}" ]; then \
	  if [ -w `dirname $(sysconfdir)` ]; then \
	    mkdir -p $(sysconfdir); \
	    if [ -w $(sysconfdir) ]; then \
	      if [ ! -f $(sysconfdir)/ftpd.conf ]; then \
	        install -c -m 0644 etc/ftpd.conf $(sysconfdir)/ftpd.conf; \
	        echo "Installing $(sysconfdir)/ftpd.conf"; \
	      else \
	        install -c -m 0644 etc/ftpd.conf $(sysconfdir)/ftpd.conf.dist; \
	        echo "Old configuration file $(sysconfdir)/ftpd.conf not overwritten."; \
	        echo "New configuration file installed as $(sysconfdir)/ftpd.conf.dist"; \
	      fi; \
	    fi; \
	  else \
	    echo "notice: ftpd.conf was not installed because you don\'t have write permission to $(sysconfdir)."; \
	  fi; \
	fi
	@echo
	@echo "Net::FTPServer has been installed. Before proceeding please"
	@echo "read the manual page. Type:"
	@echo
	@echo "  man Net::FTPServer"
	@echo
';
  }

__END__
